core/perf_test/test_atomic_minmax_simple.cpp

```diff
@@ -21,240 +21,553 @@
+        if (i == con_length - 1) {
+          Kokkos::atomic_max_fetch(&(input(0)), max);
+          inner = max;
+        }
+      },
+      Kokkos::Max<T>(current));
+  Kokkos::fence();
+  const auto time = timer.seconds();
+
+  if (current < max) {
+    report_errors_contentious_replacement(state, current, input(0), max);
   }
 
-  // input is max values - some min atomics will replace
-  {
-    Kokkos::parallel_for(
-        1, KOKKOS_LAMBDA(const int i) { inp(i) = max; });
-    Kokkos::fence();
-
-    timer.reset();
-    T current(100000000);
-    Kokkos::parallel_reduce(
-        con_length,
-        KOKKOS_LAMBDA(const int i, T& inner) {
-          inner = Kokkos::atomic_min_fetch(&(inp(0)), inner - 1);
-          if (i == con_length - 1) {
-            Kokkos::atomic_min_fetch(&(inp(0)), min);
-            inner = min;
-          }
-        },
-        Kokkos::Min<T>(current));
-    Kokkos::fence();
-    double time = timer.seconds();
-
-    if (current > min) {
-      std::cerr << "Error in contentious min replacements: " << std::endl;
-      std::cerr << "final=" << current << " inp(0)=" << inp(0) << " min=" << min
-                << std::endl;
-    }
-    std::cout << "Time for contentious min " << con_length
-              << " replacements: " << time << std::endl;
+  return time;
+}
+
+template <typename T>
+static void Atomic_ContentiousMaxReplacements(benchmark::State& state) {
+  const auto length = get_length(state);
+  auto inp          = prepare_input(1, std::numeric_limits<T>::lowest());
+
+  for (auto _ : state) {
+    const auto time = atomic_contentious_max_replacement(state, inp, length);
+
+    state.SetIterationTime(time);
+  }
+}
+
+template <typename T>
+double atomic_contentious_min_replacement(benchmark::State& state,
+                                          Kokkos::View<T*, exec_space> input,
+                                          const int con_length) {
+  const auto min = std::numeric_limits<T>::lowest();
+  T current      = 0;
+
+  Kokkos::Timer timer;
+  Kokkos::parallel_reduce(
+      con_length,
+      KOKKOS_LAMBDA(const int i, T& inner) {
+        inner = Kokkos::atomic_min_fetch(&(input(0)), inner - 1);
+        if (i == con_length - 1) {
+          Kokkos::atomic_min_fetch(&(input(0)), min);
+          inner = min;
+        }
+      },
+      Kokkos::Min<T>(current));
+  Kokkos::fence();
+  const auto time = timer.seconds();
+
+  if (current > min) {
+    report_errors_contentious_replacement(state, current, input(0), min);
+  }
+
+  return time;
+}
+
+template <typename T>
+static void Atomic_ContentiousMinReplacements(benchmark::State& state) {
+  const auto length = get_length(state);
+  auto inp          = prepare_input(1, std::numeric_limits<T>::max());
+
+  for (auto _ : state) {
+    const auto time = atomic_contentious_max_replacement(state, inp, length);
+
+    state.SetIterationTime(time);
   }
 }
 
+// int
+BENCHMARK(Atomic_MinReplacements<int>)
```

Comment:
The goal here was to do single benchmark per single scenario. Turns out to be very verbose, but I couldn't find a way to do it in a nice, programmatic way with `benchmark` provided utilities.

Other options:
- use macro to avoid repetition for different types
- group more scenarios into single benchmarks and adjust reporting of the results